disBatch.py

```diff
@@ -1,1524 +1,18 @@
-                else:
-                    extraArgs.extend([aName, str(v)])
-
-            subContext = SUB.Popen([DbUtilPath] + extraArgs, stdin=open(os.devnull, 'r'), stdout=open(uniqueId + '_context_wrap.out', 'w'), stderr=open(uniqueId + '_context_wrap.err', 'w'), close_fds=True)
-        else:
-            print('Run this script to add compute contexts:\n   ' + DbUtilPath)
-            subContext = None
-
-        driver = Driver(kvs, dbInfo, tasks, getattr(taskSource, 'resultkey', resultKey))
-        try:
-            while driver.is_alive():
-                if taskProcess and taskProcess.r:
-                    logger.warn('Task generator failed; forcing shutdown')
-                    sys.exit(taskProcess.r)
-                driver.join(PulseTime)
-                kvs.put('.controller', ('driver heart beat', None))
-        except Exception as e:
-            logger.exception('Watchdog')
-        finally:
-            try:
-                logger.info("Shutting down")
-                # TODO: Filter out engines with 0 slots (already shutdown)?
-                eks = []
-                for ex, e in driver.engines.items():
-                    eks.append(ex)
-                    e.stop()
-
-                logger.info("checking engines: %r", eks)
-                for attempt in [1, 2, 3]:
-                    neks = []
-                    for ek in eks:
-                        status = kvs.view('engine %d'%ek)
-                        logger.info('engine %s reports %s.', ek, status)
-                        if status != 'engine stopped':
-                            neks.append(ek)
-                    if not neks: break
-                    logger.info('Hold outs: %r', neks)
-                    eks = neks
-                    time.sleep(3)
-            except Exception as e:
-                logger.exception('During shutdown')
-            if kvsst:
-                logger.info('Shutting down KVS server.')
-                kvs.shutdown()
-            else:
-                kvs.close()
-            if args.kvsserver is True:
-                try:
-                    os.unlink(kvsinfotxt)
-                except OSError:
-                    # may have been already removed -- can happen if multiple disbatch runs in same directory
-                    pass
-            if args.web: os.unlink(urlfile)
-            if subContext:
-                killPatiently(subContext, 'Execution context')
-        if subContext and subContext.returncode:
-            print('Some engine processes failed -- please check the logs', file=sys.stderr)
-            sys.exit(1)
-
-        if args.exit_code and driver.failed:
-            print('Some tasks failed with non-zero exit codes -- please check the logs', file=sys.stderr)
-            sys.exit(1)
+dbExec = os.path.join(os.path.sep.join(disbatch.__file__.split(os.path.sep)[:-1]), 'disBatch.py')
+os.execv(sys.executable, [sys.executable, dbExec] + sys.argv[1:])
```

Comment:
I didn't know python could do process substitution like this :)

Reply:
It's a good thing it can. Using disbatch.main() ended up confusing the pickling of objects.